### PR TITLE
implement short_circuits function for ScalarUDFImpl trait

### LINKED ISSUE #10162: Adding the new API `is_short_circuits()`(default to false, when need set to true) to `ScalarUDF` and `ScalarUDFImpl` might be a good way to do this because users may want to define their own short-circuit functions.
### Is your feature request related to a problem or challenge?

As part of consolidating ScalarUDFs in https://github.com/apache/datafusion/pull/10098, @Omega359  and @haohuaijin  noted that there is still a special case for `coalesce`. Specifically, https://github.com/apache/datafusion/pull/10098#discussion_r1570245430_

### Describe the solution you'd like

              Adding the new API `is_short_circuits()`(default to false, when need set to true) to `ScalarUDF` and `ScalarUDFImpl` might be a good way to do this because users may want to define their own short-circuit functions.

So that would me add a function to `ScalarUDFImpl` like the following (and remove the check for "coalsce" in the code)

```rust
trait ScalarUDFImpl {
..
/// Returns true if some of this `exprs` subexpressions may not be evaluated
/// and thus any side effects (like divide by zero) may not be encountered
/// Setting this to true prevents certain optimizations such as common subexpression elimination
fn short_circuits() -> bool {
  false
}
```

Then return `true` in the implementation of `Coalsce`

_Originally posted by @haohuaijin in https://github.com/apache/datafusion/pull/10098#discussion_r1570245430_

### PR DESCRIPTION
**Pull request by Lordworms**
## Which issue does this PR close?

Closes #10162

### CODE REVIEW

File: datafusion/expr/src/udf.rs [hunk 1]
```diff
@@ -376,6 +376,12 @@ pub trait ScalarUDFImpl: Debug + Send + Sync {
     ) -> Result<ExprSimplifyResult> {
         Ok(ExprSimplifyResult::Original(args))
     }
+    /// Returns true if some of this `exprs` subexpressions may not be evaluated
```
Comment (liukun4515) [inline datafusion/expr/src/udf.rs:379]:
```suggestion

    /// Returns true if some of this `exprs` subexpressions may not be evaluated
```

**Comment by liukun4515** [comment]
after https://github.com/apache/datafusion/pull/10098 merge, you can rewrite this function for `Coalsce`

File: datafusion/expr/src/udf.rs (hunk 1 above)
Comment (liukun4515) [reply datafusion/expr/src/udf.rs:379]: blank line

**Comment by alamb** [comment]
> after #10098 merge, you can rewrite this function for `Coalsce`

#10098 is merged now. Thank you @Lordworms

File: datafusion/expr/src/udf.rs (hunk 1 above)
Comment (Lordworms) [reply datafusion/expr/src/udf.rs:379]: got it

**Review by alamb — APPROVED** [review]
Thank you @Lordworms  . Looks great to me

cc @Omega359

**Pull request merged**